### ISSUE TITLE
Result window is too large, from + size must be less than or equal to X but was Y

### ISSUE THREAD

**Issue by adminamgrade** [Smile-SA/elasticsuite#599]
We have :
Magento 2.1.8
ElasticSearch: 2.4.5.
"lucene_version" : "5.5.4"
Smile/Elasticsuite: 2.4.x

We get the error as below
```
[2017-10-31 02:18:06,309][WARN ][rest.suppressed          ] path: /magento2_storeview_1outlets_com_catalog_product/product/_search, params: {ind
ex=magento2_storeview_1outlets_com_catalog_product, type=product}
Failed to execute phase [query_fetch], all shards failed; shardFailures {[bsfRI3bIRNmHyba-bzzyOQ][magento2_storeview_1outlets_com_catalog_produc
t_20171006_104603][0]: RemoteTransportException[[Lee Forrester][172.30.1.173:9300][indices:data/read/search[phase/query+fetch]]]; nested: QueryP
haseExecutionException[Result window is too large, from + size must be less than or equal to: [100000] but was [1012373]. See the scroll api for
 a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level parameter.]; }
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.onFirstPhaseResult(AbstractSearchAsyncAction.java:206)
        at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:152)
        at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:46)
        at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:874)
        at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:852)
        at org.elasticsearch.transport.TransportService$4.onFailure(TransportService.java:389)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:39)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: QueryPhaseExecutionException[Result window is too large, from + size must be less than or equal to: [100000] but was [1012373]. See t
he scroll api for a more efficient way to request large data sets. This limit can be set by changing the [index.max_result_window] index level p
arameter.]
        at org.elasticsearch.search.internal.DefaultSearchContext.preProcess(DefaultSearchContext.java:212)
        at org.elasticsearch.search.query.QueryPhase.preProcess(QueryPhase.java:103)
        at org.elasticsearch.search.SearchService.createContext(SearchService.java:689)
        at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:633)
        at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:472)
        at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportA
ction.java:392)
        at org.elasticsearch.search.action.SearchServiceTransportAction$SearchQueryFetchTransportHandler.messageReceived(SearchServiceTransportAction.java:389)
        at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
        at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:77)
        at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:378)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
        ... 3 more

```
===>
We know that there is a quick fix is to increase index.max_result_window > X, but it is not a good idea due to CPU, RAM,... resource (very high load when we increase max_result_window)

======>
What we should do now ?, is there any way to use Scroll API with Smile Elasticsuite ?

**Comment by afoucret** [comment]
How many product have you in the index ?

Maybe you should try :
- increase `index.max_result_window`
- increase the number of shards (you can do it from the admin area)

**Comment by afoucret** [comment]
The PR #601 can help by setting the max_result_window in function of the number of shards.
So increasing the number of shard may ne enough ...

Can you check the resource consumption ?

Warning : you have to reindex the catalog after installing.

**Comment by ngocdb** [comment]
Hi, we have a similar issue.

Currently we set the index.max_result_window is 100000 too (in Elasticsearch yml configuration). But of course it is not enough because we have more than 2 millions products in a store (so an index), so we always get the error "Result window is too large, from + size must be less than or equal to: [100000] but was [2xxxxxx]"

===>

> So increasing the number of shard may ne enough ...

=>

@afoucret , so what number of shard do you suggest ?, if for an index we have more than 2 millions documents.

**Comment by afoucret** [comment]
My patch allow EACH shard to have 10000 result window.

So at max you will have 2000000 /  10000 hits = 200 shards.

But i don't think there is a lot of queries that match the whole catalog ... So having 20 shards should be enough. Maybe we can have an hybrid approach : increasing the number of shard AND the max result window (to be configurable) at the same time.

A good config may be :
* 20 shards
* 100 000 for the max_result_window

BR,

**Comment by afoucret** [comment]
Previous PR #601 have been replaced by PR #663.

It allows to have a max_window_result depending on the number of shards.
So increasing the number of shards should solve this bug in the future.

Will be available in the next stable release (2.5.0)

**Comment by davidandersson1** [comment]
Hello,

We have also same error and its continuous store in magento/var/log/system.log + elasticsearch.log in Elasticsearch server.

We have around 400k+ products with 2 store views and around 650 categories. and we are using separate instance for Elasticsearch.

Magento version 2.3.5 + Smile elastic suite module version ~2.9.0

On admin configuration,
1) Number of Shards per Index = 1
2) Number of Replicas per Index = 0

Number of document in elastic index = "2383824" according to look indices on admin->ElasticSuite->System->Indices.

@afoucret  , can you please suggest what number of shard we have to configure in admin area to fix the error?

Please help!

Thanks in advance,
David Andersson
![Screenshot from 2020-10-15 15-02-21](https://user-images.githubusercontent.com/49389320/96105291-7d7d3280-0ef7-11eb-8720-aec9365df72f.png)
